plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Utils.hs

```diff
@@ -311,38 +315,118 @@ costIsAcceptable = \case
+            , _stSeenParams = mempty
+            }
+        )
+        . go
+  where
+    go :: Term tyname name uni fun ann -> State (St name) Bool
+    go term = ifM (use stRemainingSize <&> (< 0)) (pure False) $ do
+        unseen <- use stUnseenParams
+        seen <- use stSeenParams
+        case term of
+            Builtin{} -> decSize $> True
+            Var _ n
+                | n `Set.member` seen -> pure False
+                | n `Set.member` unseen -> do
+                    modify' $ \st ->
+                        st
+                            & stUnseenParams %~ Set.delete n
+                            & stSeenParams %~ Set.insert n
+                    decSize $> True
+                | otherwise -> decSize $> True
+            Error{} -> decSize $> True
+            LamAbs _ _ _ body -> decSize >> go body
+            TyAbs _ _ _ body ->
+                -- `decSize` to account for the `TyAbs` node (which becomes `Delay` in UPLC).
+                decSize >> go body
+            Constr _ _ _ es ->
+                -- `decSize` twice to account for the `Constr` node and the index.
+                decSize >> decSize >> allM go es
+            Case _ _ arg cs -> decSize >> allM go (arg : cs)
+            IWrap _ _ _ body ->
+                -- No `decSize` here, because `IWrap` nodes disappear in UPLC.
+                go body
+            Unwrap _ body ->
+                -- No `decSize` here, because `Unwrap` nodes disappear in UPLC.
+                go body
+            -- Constants can be big! We could check the size here and inline if they're
+            -- small, but probably not worth it
+            Constant{} -> pure False
+            Apply _ fun arg ->
+                -- `decSize` to account for the `Apply` node.
+                decSize >> (go fun &&^ go arg)
+            TyInst _ body _ty ->
+                -- `decSize` to account for the `TyInst` node (which becomes `Force` in UPLC).
+                decSize >> go body
+            Let{} -> pure False
+
+    decSize = modify' (over stRemainingSize (\x -> x - 1))
 
 -- | Is this an utterly trivial type which might as well be inlined?
 trivialType :: Type tyname uni ann -> Bool
 trivialType = \case
     TyBuiltin{} -> True
     TyVar{}     -> True
     _           -> False
+
+{- Note [Size criteria for inlining]
+Inlining replaces LHS with RHS, for some `LHS = RHS`. The LHS is usually a variable and the
+RHS is usually the definition of the variable. But when we are considering inlining
+fully-saturated applications, LHS can have parameters.
+
+For example, given `x = \a b c d -> f y`, normally LHS is `x`. But when we are considering
+inlining a call site of `x` where `x` is fully applied, LHS becomes `x a b c d`, and
+RHS becomes `f y`.
+
+In order to avoid increasing program sizes (which is much more important for Plutus Core than
+it is for general-purpose programming languages), we should inline only if size(RHS) is
+no bigger than size(LHS).
```

Comment:
Do we also want to add an non-conservative mode that relaxes this a bit? I think we probably do. Maybe not here, I guess.